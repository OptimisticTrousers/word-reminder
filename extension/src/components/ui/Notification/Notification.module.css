.notification {
  border-radius: 6px;
  padding: 1rem;
}

.notification--success {
  --bkg-color: rgb(220, 252, 231);
  --base-text-color: rgb(20, 83, 45);
  --btn-text-color: rgb(22, 163, 74);
  --icon-color: rgb(21, 128, 61);
  background-color: var(--bkg-color);
}

.notification--error {
}

.notification__flex {
  display: flex;
  align-items: center;
}

.notification__item {
  flex-shrink: 0;
}

.notification__box {
  margin-left: 0.75rem;
}

.notification__message {
  color: var(--base-text-color);
  font-weight: 500;
  font-size: 0.875rem;
  line-height: 1.25rem;
  display: block;
}

.notification__controls {
  margin-left: auto;
  padding-left: 0.75rem;
}

.notification__control {
  margin-top: -0.375rem;
  margin-right: -0.375rem;
  margin-bottom: -0.375rem;
  margin-left: -0.375rem;
}

.notification__button {
  color: var(--btn-text-color);
  padding: 0.375rem;
  background-color: var(--bkg-color);
  border-radius: 6px;
  display: inline-flex;
}

.notification__text {
  position: absolute;
  width: 1px;
  padding: 1px;
  padding: 0;
  margin: -1px;
  overflow: hidden;
  clip: rect(0, 0, 0, 0);
  white-space: nowrap;
  border-width: 0;
}

.notification__icon {
  width: 1.25rem;
  height: 1.25rem;
}

.notification__icon--circle {
  color: var(--icon-color);
}
