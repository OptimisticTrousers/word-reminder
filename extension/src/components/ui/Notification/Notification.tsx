import { CircleCheck, CircleX, X } from "lucide-react";
import { useContext } from "react";
import CSSModules from "react-css-modules";

import {
  NOTIFICATION_ACTIONS,
  NotificationContext,
} from "../../../context/Notification";
import styles from "./Notification.module.css";

export const Notification = CSSModules(
  function ({ type, message }: { type: string; message: string }) {
    const { dismissNotification } = useContext(NotificationContext);

    function handleClick() {
      dismissNotification();
    }

    return (
      <div
        styleName={`notification notification--${type}`}
        role="dialog"
        aria-labelledby="notification-message"
      >
        <div styleName="notification__flex">
          <div styleName="notification__item">
            {type === NOTIFICATION_ACTIONS.SUCCESS ? (
              <CircleCheck
                data-testid="check-icon"
                styleName="notification__icon notification__icon--circle"
              />
            ) : (
              <CircleX
                data-testid="x-icon"
                styleName="notification__icon notification__icon--circle"
              />
            )}
          </div>
          <div styleName="notification__box">
            <span id="notification-message" styleName="notification__message">
              {message}
            </span>
          </div>
          <div styleName="notification__controls">
            <div styleName="notification__control">
              <button styleName="notification_button" onClick={handleClick}>
                <span styleName="notification__text">Dismiss</span>
                <X styleName="notification__icon" />
              </button>
            </div>
          </div>
        </div>
      </div>
    );
  },
  styles,
  { allowMultiple: true, handleNotFoundStyleName: "log" }
);
