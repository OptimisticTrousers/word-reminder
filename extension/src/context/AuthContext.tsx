import { createContext, FC, useReducer } from "react";

interface IUser {
  _id: string;
  username: string;
  theme: string;
  words: IUserWord[];
}

interface IWord {
  _id: string;
  word: string;
  origin: string;
  phonetic: string;
  audio: string;
  meanings: [
    {
      partOfSpeech: string;
      definitions: [
        {
          definition: string;
          example: string;
          synonyms: string[];
          antonyms: string[];
        }
      ];
    }
  ];
}

export interface IUserWord {
  _id: string;
  userId: IUser | string;
  word: IWord;
  learned: boolean;
}

interface AuthState {
  user: IUser | null;
}

interface Action {
  type: string;
  payload: IUser;
}

interface AuthContext {
  state: AuthState;
  dispatch: (action: Action) => void;
}

const authReducer = (state: AuthState, action: Action) => {
  switch (action.type) {
    // Login action will always contain a payload of user object
    case "LOGIN":
      return { user: action.payload };
    // Remove user object on logout action dispatch
    case "LOGOUT":
      return { user: null };
    default:
      return state;
  }
};

export const AuthContext = createContext<AuthContext>({} as AuthContext);

interface Props {
  children: JSX.Element[] | JSX.Element;
}

const AuthProvider: FC<Props> = ({ children }) => {
  const [state, dispatch] = useReducer(authReducer, {
    user: null,
  });

  return (
    <AuthContext.Provider value={{ state, dispatch }}>
      {children}
    </AuthContext.Provider>
  );
};

export default AuthProvider;
