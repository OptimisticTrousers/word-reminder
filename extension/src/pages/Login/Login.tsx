/* eslint-disable @typescript-eslint/no-explicit-any */
import CSSModules from "react-css-modules";
import styles from "../../assets/Auth.module.css";
import { useForm } from "react-hook-form";
import { zodResolver } from "@hookform/resolvers/zod";
import * as z from "zod";
import { Link, useNavigate } from "react-router-dom";
import { useMutation } from "@tanstack/react-query";
import { toast } from "react-toastify";

const schema = z.object({
  username: z.string({
    required_error: "Username is required",
    invalid_type_error: "Username must be a string",
  }),
  password: z.string({
    required_error: "Password is required",
    invalid_type_error: "Password must be a string",
  }),
});

const Login = CSSModules(
  () => {
    const {
      register,
      handleSubmit,
      formState: { errors },
    } = useForm({
      defaultValues: {
        username: "",
        password: "",
      },
      resolver: zodResolver(schema),
    });

    const navigate = useNavigate();

    const { data, status, error, mutate }: any = useMutation({
      mutationFn: (formData) => {
        return fetch(`${import.meta.env.VITE_API_DOMAIN}/auth/login`, {
          body: JSON.stringify(formData),
        });
      },
      onSuccess: () => {
        toast.success("You have successfully logged in!");
        navigate("/");
      },
      onError: () => {
        toast.error("There was an issue logging in!");
      },
    });

    console.log(data);
    console.log(status);
    console.log(error);

    const onSubmit = handleSubmit(mutate);

    return (
      <section styleName="auth auth--login">
        <div styleName="auth__box">
          <div styleName="auth__hero">
            <h2>Word Storer</h2>
            <p styleName="auth__description">
              Log in or create a new account to start storing your words.
            </p>
          </div>
        </div>
        <form styleName="auth__form" onSubmit={onSubmit}>
          <div styleName="auth__control">
            <label htmlFor="username" styleName="auth__label">
              <span styleName="auth__bold">Username</span>
            </label>
            <input
              {...register("username", { required: "This is required." })}
              styleName="auth__input"
            />
            <p styleName="auth__error">{errors.username?.message}</p>
          </div>
          <div styleName="auth__control">
            <label htmlFor="password" styleName="auth__label">
              <span styleName="auth__bold">Password</span>
            </label>
            <input
              type="password"
              {...register("password", { required: "This is required." })}
              styleName="auth__input"
            />
            <p styleName="auth__error">{errors.username?.message}</p>
          </div>
          <button styleName="auth__button" type="submit">
            Login
          </button>
        </form>
        <p styleName="auth__create">
          New around here? <Link to="/register">Create account</Link>
        </p>
      </section>
    );
  },
  styles,
  { allowMultiple: true, handleNotFoundStyleName: "ignore" }
);

export default Login;
