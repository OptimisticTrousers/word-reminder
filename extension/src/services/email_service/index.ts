export { emailService } from "./email_service";
