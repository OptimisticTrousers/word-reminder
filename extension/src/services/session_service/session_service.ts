import { service } from "../service";

export const sessionService = (function (service) {
  const { get, post, remove, VITE_API_DOMAIN } = service;

  function getCurrentUser() {
    return get({
      url: `${VITE_API_DOMAIN}/sessions`,
      options: {
        credentials: "include",
      },
    });
  }

  function logoutUser() {
    return remove({
      url: `${VITE_API_DOMAIN}/sessions`,
      options: {
        credentials: "include",
      },
    });
  }

  function loginUser(body: { email: string; password: string }) {
    return post({
      url: `${VITE_API_DOMAIN}/sessions`,
      options: {
        body: JSON.stringify(body),
        credentials: "include",
        headers: {
          "Content-Type": "application/json",
        },
      },
    });
  }

  return { getCurrentUser, logoutUser, loginUser };
})(service);
