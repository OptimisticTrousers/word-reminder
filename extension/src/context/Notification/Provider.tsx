import { useReducer, ReactNode } from "react";

import { NOTIFICATION_ACTIONS, NotificationContext } from "./Context";
import { Notification } from "../../components/ui/Notification/Notification";

interface Props {
  children: ReactNode;
}

interface NotificationState {
  isVisible: boolean;
  type: string;
  message: string;
}

interface NotificationAction {
  payload: { message: string };
  type: string;
}

export interface NotificationContext {
  showNotification: (type: string, message: string) => void;
  dismissNotification: () => void;
}

const notificationReducer = (
  state: NotificationState,
  action: NotificationAction
) => {
  switch (action.type) {
    case NOTIFICATION_ACTIONS.SUCCESS:
      return {
        ...state,
        isVisible: true,
        type: NOTIFICATION_ACTIONS.SUCCESS,
        message: action.payload.message,
      };
    case NOTIFICATION_ACTIONS.ERROR:
      return {
        ...state,
        isVisible: true,
        type: NOTIFICATION_ACTIONS.ERROR,
        message: action.payload.message,
      };
    case DISAPPEAR:
      return { ...state, isVisible: false, type: "", message: "" };
    default: {
      throw Error("Unknown action: " + action.type);
    }
  }
};

// Provide ability to spawn Toast notification globally within the application
export function NotificationProvider({ children }: Props) {
  const [state, dispatch] = useReducer(notificationReducer, {
    isVisible: false,
    type: "",
    message: "",
  });

  function dismissNotification() {
    dispatch({
      type: DISAPPEAR,
      payload: { message: "" },
    });
  }

  function showNotification(type: string, message: string): void {
    const delay = 5000;

    dispatch({ type, payload: { message } });

    setTimeout(() => {
      dismissNotification();
    }, delay);
  }

  const { isVisible, type, message } = state;

  return (
    <NotificationContext.Provider
      value={{ showNotification, dismissNotification }}
    >
      {isVisible && <Notification type={type} message={message} />}
      {children}
    </NotificationContext.Provider>
  );
}

const DISAPPEAR = "disappear";
