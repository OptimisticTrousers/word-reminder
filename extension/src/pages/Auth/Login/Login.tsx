import { EMAIL_MAX, PASSWORD_MAX, User } from "common";
import { useContext } from "react";
import CSSModules from "react-css-modules";
import { Link } from "react-router-dom";
import { useMutation, useQueryClient } from "@tanstack/react-query";

import { AUTH_NOTIFICATION_MSGS } from "../constants";
import {
  NOTIFICATION_ACTIONS,
  NotificationContext,
} from "../../../context/Notification";
import { sessionService } from "../../../services/session_service";
import { Response } from "../../../types";
import styles from "../Auth.module.css";
import { useNotificationError } from "../../../hooks/useNotificationError";

export const Login = CSSModules(
  function () {
    const { showNotification } = useContext(NotificationContext);
    const { showNotificationError } = useNotificationError();
    const queryClient = useQueryClient();

    const { status, mutate } = useMutation({
      mutationFn: sessionService.loginUser,
      onSuccess: (response: Response & { json: { user: User } }) => {
        showNotification(
          NOTIFICATION_ACTIONS.SUCCESS,
          LOGIN_AUTH_NOTIFICATION_MSGS.login(response.json.user.email)
        );
        queryClient.invalidateQueries({ queryKey: ["sessions"], exact: true });
      },
      onError: showNotificationError,
    });

    function handleSubmit(formData: FormData) {
      mutate({
        email: formData.get("email") as string,
        password: formData.get("password") as string,
      });
    }

    const disabled = status === "pending";

    return (
      <section styleName="auth auth--login">
        <div styleName="auth__hero">
          <h2 styleName="auth__heading">Word Storer</h2>
          <p styleName="auth__description">
            Log in or create a new account to start storing your words.
          </p>
        </div>
        <form styleName="auth__form" action={handleSubmit}>
          <div styleName="auth__control">
            <label htmlFor="email" styleName="auth__label">
              Email <span styleName="auth__required">(required)</span>
            </label>
            <input
              type="email"
              id="email"
              name="email"
              styleName="auth__input"
              required={true}
              disabled={disabled}
              maxLength={EMAIL_MAX}
            />
          </div>
          <div styleName="auth__control">
            <label htmlFor="password" styleName="auth__label">
              Password <span styleName="auth__required">(required)</span>
            </label>
            <input
              type="password"
              id="password"
              name="password"
              styleName="auth__input"
              required={true}
              disabled={disabled}
              maxLength={PASSWORD_MAX}
            />
          </div>
          <button styleName="auth__button" type="submit" disabled={disabled}>
            Login
          </button>
        </form>
        <p styleName="auth__redirect">
          New around here?{" "}
          <Link styleName="auth_link" to="/register">
            Create account
          </Link>
        </p>
        <p styleName="auth__redirect">
          <Link styleName="auth_link" to="/forgotPassword">
            Forgot Password?
          </Link>
        </p>
      </section>
    );
  },
  styles,
  { allowMultiple: true, handleNotFoundStyleName: "log" }
);

const LOGIN_AUTH_NOTIFICATION_MSGS = {
  ...AUTH_NOTIFICATION_MSGS,
  login: (email: string) => {
    return `You have successfully logged in, ${email}.`;
  },
};
