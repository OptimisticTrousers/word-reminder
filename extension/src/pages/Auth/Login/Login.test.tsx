import { EMAIL_MAX, PASSWORD_MAX } from "common";
import { createRoutesStub } from "react-router-dom";
import { QueryClient, QueryClientProvider } from "@tanstack/react-query";
import { render, screen } from "@testing-library/react";
import userEvent from "@testing-library/user-event";

import { NotificationProvider } from "../../../context/Notification";
import { Login } from "./Login";
import { sessionService } from "../../../services/session_service";

describe("Login component", () => {
  const Stub = createRoutesStub([
    {
      path: "/login",
      Component: Login,
    },
    {
      path: "/words",
      Component: function () {
        return <button>Current URL: /words</button>;
      },
    },
  ]);

  const queryClient = new QueryClient({
    defaultOptions: {
      queries: {
        retry: false,
      },
    },
  });

  const testUser = {
    email: "bob@gmail.com",
    password: "password",
  };
  const status = 200;

  function setup() {
    return {
      user: userEvent.setup(),
      ...render(
        <NotificationProvider>
          <QueryClientProvider client={queryClient}>
            <Stub initialEntries={["/login"]} />
          </QueryClientProvider>
        </NotificationProvider>
      ),
    };
  }

  beforeEach(() => {
    vi.clearAllMocks();
  });

  it("calls the functions to login, show notification, and redirect user", async () => {
    const mockSessionServiceLogin = vi
      .spyOn(sessionService, "loginUser")
      .mockImplementation(async () => {
        return { json: { user: testUser }, status };
      });
    const { asFragment, user } = setup();
    const email = testUser.email;
    const password = testUser.password;
    const passwordInput = screen.getByLabelText("Password (required)", {
      selector: "input",
    });
    const emailInput = screen.getByLabelText("Email (required)", {
      selector: "input",
    });
    await user.type(emailInput, email);
    await user.type(passwordInput, password);

    const loginButton = screen.getByRole("button", { name: "Login" });
    await user.click(loginButton);

    const words = screen.getByRole("button", { name: "Current URL: /words" });
    const notification = screen.getByRole("dialog", {
      name: `You have successfully logged in, ${email}.`,
    });
    expect(mockSessionServiceLogin).toHaveBeenCalledTimes(1);
    expect(mockSessionServiceLogin).toHaveBeenCalledWith(testUser);
    expect(words).toBeInTheDocument();
    expect(notification).toBeInTheDocument();
    expect(asFragment()).toMatchSnapshot();
  });

  it("shows correct link to creating an account", async () => {
    setup();
    const link = screen.getByRole("link", { name: "Create account" });

    expect(link).toHaveAttribute("href", "/register");
  });

  describe("form validation", () => {
    it("allows user to type on the email field", async () => {
      const { user } = setup();
      const email = testUser.email;
      const emailInput = screen.getByLabelText("Email (required)", {
        selector: "input",
      });

      await user.type(emailInput, email);

      expect(emailInput).toHaveValue(email);
    });

    it("allows user to type on the password field", async () => {
      const { user } = setup();
      const password = testUser.password;
      const passwordInput = screen.getByLabelText("Password (required)", {
        selector: "input",
      });

      await user.type(passwordInput, password);

      expect(passwordInput).toHaveValue(password);
    });

    it(`does not allow user to type more than ${EMAIL_MAX} characters in email field`, async () => {
      const { user } = setup();
      const email = new Array(EMAIL_MAX + 1).fill("a").join("");
      const emailInput = screen.getByLabelText("Email (required)", {
        selector: "input",
      });

      await user.type(emailInput, email);

      expect(emailInput).toHaveValue(email.slice(0, -1));
    });

    it(`does not allow user to type more than ${PASSWORD_MAX} characters in password field`, async () => {
      const { user } = setup();
      const password = new Array(PASSWORD_MAX + 1).fill("a").join("");
      const passwordInput = screen.getByLabelText("Password (required)", {
        selector: "input",
      });

      await user.type(passwordInput, password);

      expect(passwordInput).toHaveValue(password.slice(0, -1));
    });

    it("does not allow the user to login when they provided an email but not a password", async () => {
      const mockSessionServiceLogin = vi
        .spyOn(sessionService, "loginUser")
        .mockImplementation(async () => {
          return { json: { user: testUser }, status };
        });
      const { user } = setup();
      const email = testUser.email;
      const emailInput = screen.getByLabelText("Email (required)", {
        selector: "input",
      });
      await user.type(emailInput, email);

      const loginButton = screen.getByRole("button", { name: "Login" });
      await user.click(loginButton);

      expect(mockSessionServiceLogin).not.toHaveBeenCalled();
    });

    it("does not allow the user to login when they provided a password but not an email", async () => {
      const mockSessionServiceLogin = vi
        .spyOn(sessionService, "loginUser")
        .mockImplementation(async () => {
          return { json: { user: testUser }, status };
        });
      const { user } = setup();
      const password = testUser.password;
      const passwordInput = screen.getByLabelText("Password (required)", {
        selector: "input",
      });
      await user.type(passwordInput, password);

      const loginButton = screen.getByRole("button", { name: "Login" });
      await user.click(loginButton);

      expect(mockSessionServiceLogin).not.toHaveBeenCalled();
    });

    it("does not allow the user to login when they did not provide a password or email", async () => {
      const mockSessionServiceLogin = vi
        .spyOn(sessionService, "loginUser")
        .mockImplementation(async () => {
          return { json: { user: testUser }, status };
        });
      const { user } = setup();

      const loginButton = screen.getByRole("button", { name: "Login" });
      await user.click(loginButton);

      expect(mockSessionServiceLogin).not.toHaveBeenCalled();
    });

    it("shows a notification when there is an error", async () => {
      const message = "Error";
      const mockSessionServiceLogin = vi
        .spyOn(sessionService, "loginUser")
        .mockImplementation(async () => {
          return Promise.reject({ json: { message }, status: 400 });
        });
      const { user } = setup();
      const email = testUser.email;
      const password = testUser.password;
      const passwordInput = screen.getByLabelText("Password (required)", {
        selector: "input",
      });
      const emailInput = screen.getByLabelText("Email (required)", {
        selector: "input",
      });
      await user.type(emailInput, email);
      await user.type(passwordInput, password);

      const loginButton = screen.getByRole("button", { name: "Login" });
      await user.click(loginButton);

      const notification = screen.getByRole("dialog", {
        name: message,
      });
      expect(notification).toBeInTheDocument();
      expect(mockSessionServiceLogin).toHaveBeenCalledTimes(1);
      expect(mockSessionServiceLogin).toHaveBeenCalledWith(testUser);
    });

    it("disables the form elements and buttons when while it is loading", async () => {
      const delay = 500;
      vi.spyOn(sessionService, "loginUser").mockImplementation(() => {
        return new Promise((resolve) => {
          setTimeout(() => {
            resolve({ json: { user: testUser }, status });
          }, delay);
        });
      });
      const { user } = setup();
      const email = testUser.email;
      const password = testUser.password;
      const passwordInput = screen.getByLabelText("Password (required)", {
        selector: "input",
      });
      const emailInput = screen.getByLabelText("Email (required)", {
        selector: "input",
      });
      await user.type(emailInput, email);
      await user.type(passwordInput, password);

      const loginButton = screen.getByRole("button", { name: "Login" });
      await user.click(loginButton);

      expect(emailInput).toBeDisabled();
      expect(passwordInput).toBeDisabled();
      expect(loginButton).toBeDisabled();
    });
  });
});
