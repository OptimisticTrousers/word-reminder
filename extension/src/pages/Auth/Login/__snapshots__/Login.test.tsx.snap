// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Login component > calls the functions to login, show notification, and redirect user 1`] = `
<DocumentFragment>
  <div
    aria-labelledby="notification-message"
    class="_notification_120862 _notification--success_120862"
    role="dialog"
  >
    <div
      class="_notification__flex_120862"
    >
      <div
        class="_notification__item_120862"
      >
        <svg
          class="lucide lucide-circle-check _notification__icon_120862 _notification__icon--circle_120862"
          data-testid="check-icon"
          fill="none"
          height="24"
          stroke="currentColor"
          stroke-linecap="round"
          stroke-linejoin="round"
          stroke-width="2"
          viewBox="0 0 24 24"
          width="24"
          xmlns="http://www.w3.org/2000/svg"
        >
          <circle
            cx="12"
            cy="12"
            r="10"
          />
          <path
            d="m9 12 2 2 4-4"
          />
        </svg>
      </div>
      <div
        class="_notification__box_120862"
      >
        <span
          class="_notification__message_120862"
          id="notification-message"
        >
          You have successfully logged in, bob@gmail.com.
        </span>
      </div>
      <div
        class="_notification__controls_120862"
      >
        <div
          class="_notification__control_120862"
        >
          <button
            class="_notification_button_120862"
          >
            <span
              class="_notification__text_120862"
            >
              Dismiss
            </span>
            <svg
              class="lucide lucide-x _notification__icon_120862"
              fill="none"
              height="24"
              stroke="currentColor"
              stroke-linecap="round"
              stroke-linejoin="round"
              stroke-width="2"
              viewBox="0 0 24 24"
              width="24"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M18 6 6 18"
              />
              <path
                d="m6 6 12 12"
              />
            </svg>
          </button>
        </div>
      </div>
    </div>
  </div>
  <section
    class="_auth_340faf _auth--login_340faf"
  >
    <div
      class="_auth__hero_340faf"
    >
      <h2
        class="_auth__heading_340faf"
      >
        Word Storer
      </h2>
      <p
        class="_auth__description_340faf"
      >
        Log in or create a new account to start storing your words.
      </p>
    </div>
    <form
      action="javascript:throw new Error('A React form was unexpectedly submitted. If you called form.submit() manually, consider using form.requestSubmit() instead. If you\\'re trying to use event.stopPropagation() in a submit event handler, consider also calling event.preventDefault().')"
      class="_auth__form_340faf"
    >
      <div
        class="_auth__control_340faf"
      >
        <label
          class="_auth__label_340faf"
          for="email"
        >
          Email 
          <span
            class="_auth__required_340faf"
          >
            (required)
          </span>
        </label>
        <input
          class="_auth__input_340faf"
          id="email"
          maxlength="255"
          name="email"
          required=""
          type="email"
        />
      </div>
      <div
        class="_auth__control_340faf"
      >
        <label
          class="_auth__label_340faf"
          for="password"
        >
          Password 
          <span
            class="_auth__required_340faf"
          >
            (required)
          </span>
        </label>
        <input
          class="_auth__input_340faf"
          id="password"
          maxlength="72"
          name="password"
          required=""
          type="password"
        />
      </div>
      <button
        class="_auth__button_340faf"
        type="submit"
      >
        Login
      </button>
    </form>
    <p
      class="_auth__redirect_340faf"
    >
      New around here? 
      <a
        class="_auth_link_340faf"
        data-discover="true"
        href="/register"
      >
        Create account
      </a>
    </p>
    <p
      class="_auth__redirect_340faf"
    >
      <a
        class="_auth_link_340faf"
        data-discover="true"
        href="/forgotPassword"
      >
        Forgot Password?
      </a>
    </p>
  </section>
</DocumentFragment>
`;
